### PR TITLE
Fix freeing revocable memory after query moved to reserved pool

### LINKED ISSUE #9488: Getting "Tried to free more revocable memory than is reserved" randomly when using spill
Presto Version - 0.188
Here is stacktrace for the error.

```
java.lang.IllegalArgumentException: tried to free more revocable memory than is reserved
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
        at com.facebook.presto.memory.MemoryPool.freeRevocable(MemoryPool.java:196)
        at com.facebook.presto.memory.QueryContext.freeRevocableMemory(QueryContext.java:168)
        at com.facebook.presto.operator.TaskContext.freeRevocableMemory(TaskContext.java:226)
        at com.facebook.presto.operator.PipelineContext.freeRevocableMemory(PipelineContext.java:274)
        at com.facebook.presto.operator.DriverContext.freeRevocableMemory(DriverContext.java:254)
        at com.facebook.presto.operator.OperatorContext.freeRevocableMemory(OperatorContext.java:315)
        at com.facebook.presto.operator.OperatorContext.setRevocableMemoryReservation(OperatorContext.java:307)
        at com.facebook.presto.operator.HashBuilderOperator.lambda$close$11(HashBuilderOperator.java:644)
        at com.google.common.io.Closer.close(Closer.java:216)
        at com.facebook.presto.operator.HashBuilderOperator.close(HashBuilderOperator.java:645)
        at com.facebook.presto.operator.HashBuilderOperator.finish(HashBuilderOperator.java:450)
        at com.facebook.presto.operator.Driver.processInternal(Driver.java:357)
        at com.facebook.presto.operator.Driver.lambda$processFor$6(Driver.java:241)
        at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:614)
        at com.facebook.presto.operator.Driver.processFor(Driver.java:235)
        at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:622)
        at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at com.facebook.presto.execution.executor.LegacyPrioritizedSplitRunner.process(LegacyPrioritizedSplitRunner.java:23)
        at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by findepi**
Fixes #9488

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 1]
```diff
@@ -186,32 +186,23 @@ public synchronized void freeSpill(long bytes)
...
+
+        ListenableFuture<?> memoryFuture = pool.reserve(queryId, reserved);
+        originalPool.free(queryId, reserved);
+
+        ListenableFuture<?> revocableMemoryFuture = pool.reserveRevocable(queryId, revocableReserved);
+        originalPool.freeRevocable(queryId, revocableReserved);
+
+        Futures.allAsList(memoryFuture, revocableMemoryFuture).addListener(() -> {
+            // Unblock all the tasks, if they were waiting for memory, since we're in a new pool.
+            taskContexts.values().forEach(TaskContext::moreMemoryAvailable);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: This listener and `TaskContext::moreMemoryAvailable` will be called when *both* `memoryFuture` and `revocableMemoryFuture` are completed. And `TaskContext::moreMemoryAvailable` eventually completes `OperatorContext.memoryFuture`, which is for the user memory. There is also a separate `revocableMemoryFuture` in `OperatorContext` for revocable memory. So why do we need the `revocableMemoryFuture` to complete for completing `OperatorContext.memoryFuture`, which is the future for user memory?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: You're right, fixing

**Comment by findepi** [comment]
@nezihyigitbasi rebased on top of `Refactor memory tracking` #9636

File: presto-main/src/test/java/com/facebook/presto/memory/TestQueryContext.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public Object[][] testSetMemoryPoolOptions()
+    {
+        return new Object[][] {
+                {false},
+                {true},
+        };
+    }
+
+    @Test(dataProvider = "testSetMemoryPoolOptions")
+    public void testSetMemoryPool(boolean reservedPoolInitiallyUsed)
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/memory/TestQueryContext.java]: nit: `reservedPoolInitiallyUsed ` -> `useReservedPool`

File: presto-main/src/test/java/com/facebook/presto/memory/TestQueryContext.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                    new DataSize(10, BYTE),
+                    new MemoryPool(GENERAL_POOL, new DataSize(10, BYTE)),
+                    new MemoryPool(SYSTEM_POOL, new DataSize(10, BYTE)),
+                    localQueryRunner.getExecutor(),
+                    localQueryRunner.getScheduler(),
+                    new DataSize(0, BYTE),
+                    new SpillSpaceTracker(new DataSize(0, BYTE)));
+
+            // Use memory
+            LocalMemoryContext memoryContext = queryContext.getQueryMemoryContext().localUserMemoryContext();
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/memory/TestQueryContext.java]: `memoryContext` -> `userMemoryContext`

File: presto-main/src/test/java/com/facebook/presto/memory/TestQueryContext.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            assertTrue(revocableMemoryContext.setBytes(5).isDone());
+
+            queryContext.setMemoryPool(reservedPool);
+
+            if (reservedPoolInitiallyUsed) {
+                reservedPool.free(secondQuery, secondQueryMemory);
+            }
+
+            // Free memory
+            memoryContext.setBytes(0);
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/memory/TestQueryContext.java]:
```
memoryContext.close();
revocableMemoryContext.close();
```

**Comment by findepi** [comment]
AC.
I noticed that only thanks to yours https://github.com/prestodb/presto/pull/9636

**Pull request merged**